src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -90,8 +90,15 @@ RebuildQueryStrings(Query *originalQuery, List *taskList)
 			}
 		}
 
-		deparse_shard_query(query, relationId, task->anchorShardId,
-							newQueryString);
+		if (query->commandType == CMD_UPDATE)
+		{
+			pg_get_query_def(query, newQueryString);
+		}
+		else
+		{
+			deparse_shard_query(query, relationId, task->anchorShardId,
+								newQueryString);
```

Comment:
Are you proposing this to relation id problem I described above? If yes, this does not solve the problem by itself because deparse_shard_query() uses the same path. We called UpdateRelationToShardNames() through RouterSelectJob() path which already updated RTE types. This was not a problem for RouterModifyJob() which does not call UpdateRelationToShardNames() if it requires reevaluation in the executor. I did a similar hack with this commit: https://github.com/citusdata/citus/pull/1483/commits/bc04774d6e36c83e9cbd8a86176c1ea2515d850c

I will look to unify and polish this path.